src/Libraries/Microsoft.Extensions.Http.Diagnostics/Logging/Internal/HttpResponseBodyReader.cs

```diff
@@ -54,90 +51,186 @@ public ValueTask<string> ReadAsync(HttpResponseMessage response, CancellationTok
+    }
+
+    private static async ValueTask<string> ReadFromStreamWithTimeoutAsync(HttpResponseMessage response, TimeSpan readTimeout, int readSizeLimit, CancellationToken cancellationToken)
+    {
+        using var joinedTokenSource = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
+        joinedTokenSource.CancelAfter(readTimeout);
+
+        // TimeSpan.Zero cannot be set from user's code as
+        // validation prevents values less than one millisecond
+        // However, this is useful during unit tests
+        if (readTimeout <= TimeSpan.Zero)
+        {
+            // cancel immediately, async cancel not required in tests
+#pragma warning disable CA1849 // Call async methods when in an async method
+            joinedTokenSource.Cancel();
+#pragma warning restore CA1849 // Call async methods when in an async method
+        }
+
+        try
+        {
+            return await ReadFromStreamAsync(response, readSizeLimit, joinedTokenSource.Token).ConfigureAwait(false);
+        }
+
+        // when readTimeout occurred: joined token source is cancelled and cancellationToken is not
+        catch (OperationCanceledException) when (joinedTokenSource.IsCancellationRequested && !cancellationToken.IsCancellationRequested)
+        {
+            return Constants.ReadCancelled;
+        }
     }
 
-    private static async ValueTask<string> ReadFromStreamAsync(HttpResponseMessage response, int readSizeLimit,
-        RecyclableMemoryStreamManager streamManager, CancellationToken cancellationToken)
+    private static async ValueTask<string> ReadFromStreamAsync(HttpResponseMessage response, int readSizeLimit, CancellationToken cancellationToken)
     {
-#if NET5_0_OR_GREATER
+#if NET6_0_OR_GREATER
         var streamToReadFrom = await response.Content.ReadAsStreamAsync(cancellationToken).ConfigureAwait(false);
 #else
         var streamToReadFrom = await response.Content.ReadAsStreamAsync().WaitAsync(cancellationToken).ConfigureAwait(false);
 #endif
 
-        var bufferWriter = _bufferWriterPool.Get();
-        var memory = bufferWriter.GetMemory(readSizeLimit).Slice(0, readSizeLimit);
-#if !NETCOREAPP3_1_OR_GREATER
-        byte[] buffer = memory.ToArray();
-#endif
-        try
+        var pipe = new Pipe();
+
+        var bufferedString = await BufferStreamAndWriteToPipeAsync(streamToReadFrom, pipe.Writer, readSizeLimit, cancellationToken).ConfigureAwait(false);
+
+        // if stream is seekable we can just rewind it and return the buffered string
+        if (streamToReadFrom.CanSeek)
+        {
+            streamToReadFrom.Seek(0, SeekOrigin.Begin);
+
+            await pipe.Reader.CompleteAsync().ConfigureAwait(false);
+
+            return bufferedString;
+        }
+
+        // if stream is not seekable we need to write the rest of the stream to the pipe
+        // and create a new response content with the pipe reader as stream
+        _ = Task.Run(async () =>
```

Comment:
Why do we need to start a task here?

Reply:
We are calling `ConfigureAwait(false)` but don't await the task. Is that correct?

Reply:
This approach leverages `Pipe` for efficient memory management and non-blocking operations. By consuming the original stream in a background task and writing it to the Pipe, we can quickly return control to the caller. The response content is replaced with a new `StreamContent` supplied by `PipeReader.AsStream()`, allowing the application to handle large responses asynchronously and efficiently. This method improves performance and scalability by utilizing Pipe's buffer management and concurrent read/write capabilities.

Reply:
Got it, then we don't need `ConfigurateAwait(false)` there since we don't await the task, right? Or am I missing something?

Reply:
It was removed